client/src/main.rs

```diff
@@ -196,52 +195,193 @@ fn get_region_info(guest: &Arc<Guest>) -> Result<RegionInfo, CrucibleError> {
         total_size,
         total_blocks,
-        write_count,
+        write_log,
         max_block_io,
     })
 }
 
-fn update_region_info(
+/**
+ * The write log is a recording of the number of times we have written to
+ * a specific block (index in the Vec).  The write count is used to generate
+ * a known pattern to either fill the block with, or to expect from the
+ * block when reading.
+ *
+ * This is fine for an initial fill/verify framework of sorts, but there
+ * are many kinds of errors this will not find.  There are also many high
+ * performance better coverage kinds of data integrity tests, and the intent
+ * here is to balance urgency with rigor in that we can make use of external
+ * tests for the more complicated cases, and catch the easy ones here.
+ *
+ * In addition to the current write count, we make a second copy of the
+ * write count when the commit method is called.  This can be used to record
+ * the write count of a region at a specific time (like a flush) and then
+ * later used to verify that a given block has data in it from at minimum
+ * that commit, but up to the current write count.
+ *
+ * The "seed" is the current counter as a u8 for a given block.
+ */
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct WriteLog {
+    count_cur: Vec<u32>,
+    count_min: Vec<u32>,
+}
+
+impl WriteLog {
+    pub fn new(size: usize) -> Self {
+        let count_cur = vec![0_u32; size];
+        let count_min = vec![0_u32; size];
+
+        WriteLog {
+            count_cur,
+            count_min,
+        }
+    }
+
+    pub fn is_empty(&mut self) -> bool {
+        self.count_cur.is_empty()
+    }
+
+    pub fn len(&mut self) -> usize {
+        self.count_cur.len()
+    }
+
+    // If the write count is zero then we have no record of what this
+    // block contains.
+    pub fn unwritten(&self, index: usize) -> bool {
+        self.count_cur[index] == 0
+    }
+
+    // This is called when we are about to write to a block and we want
+    // to indicate that the write counter should be updated.
+    pub fn update_wc(&mut self, index: usize) {
+        assert!(self.count_cur[index] >= self.count_min[index]);
+        // TODO: handle more than u32 max writes to the same location.
+        self.count_cur[index] += 1;
+    }
+
+    // This returns the value we should expect to find at the given index,
+    // and will fit in a u8.  If we are using the seed to fill a write
+    // volume, then update_wc() should be called first to increment the
+    // counter before we get a new seed.
+    fn get_seed(&self, index: usize) -> u8 {
+        (self.count_cur[index] % 250) as u8
+    }
+
+    // This is called before a test where we expect to be recovering and we
+    // want to record the current write log values as a minimum of what
+    // we expect the counters to be.
+    pub fn commit(&mut self) {
+        // The commit needs to walk the vec and make sure any minimum
+        // values are not lost
```

Comment:
is this something that's still TODO?

Reply:
That comment is old, so it's a TOREMOVE.

Removing it before pushing.